File: CPP/ByTier/F/1759_F_All_Possible_Digits.cpp
// Problem: CF 1759 F - All Possible Digits
// https://codeforces.com/contest/1759/problem/F

/*
Code Purpose:
This program determines the minimum number of operations required to ensure that all digits from 0 to p-1 appear at least once in the number x written in base p.
Each operation increments the number x by 1, which might cause carryovers affecting the digits.

Algorithms/Techniques:
- Digit-by-digit simulation with carry handling.
- Tracking used digits using a map.
- Efficient computation of how many new digits are needed by simulating carries.
- Optimization to avoid unnecessary iterations.

Time Complexity: O(n * p) in worst case due to nested loops, but typically much better in practice.
Space Complexity: O(p) for the map storing digit usage.

*/
#include <iostream>
#include <map>

using namespace std;
int t, n, p, i, c, ans, a[110];
map<int, int> m;
int main() {
  cin >> t;
  while (t--) {
    cin >> n >> p;
    for (int i = 1; i <= n; i++)
      cin >> a[i], m[a[i]] = 1; // Read digits and mark them as present
    ans = p; // Initially assume we need p operations to cover all digits
    i = a[n]; // Start checking from the least significant digit
    // Count how many digits in the range [0, i] are missing
    for (; m[i] && i > 0; i--)
      ans--;
    if (i > 0) { // If there is a gap in used digits
      c = n - 1; // Start from second to last digit
      // Find the first digit that doesn't produce a carry when incremented
      while (c && a[c] == p - 1)
        c--;
      // If carry propagation leads back to the gap
      if ((a[c] + 1) % p == i) 
        for (i--, ans--; m[i] && i > 0; i--)
          ans--;
    } else
      // If we've covered from 0 up to i (which is 0), check backward
      for (; m[i] && ans != 0; i = (i + p - 1) % p)
        ans--;
    cout << ans << '\n';
    m.clear(); // Clear the map for next test case
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/F/1759_F_All_Possible_Digits.py
# Problem: CF 1759 F - All Possible Digits
# https://codeforces.com/contest/1759/problem/F

"""
Algorithm: Simulation with digit manipulation
Time Complexity: O(n + p) per test case, where n is the length of the number and p is the base
Space Complexity: O(p) for storing the set of seen digits

The approach simulates the process of incrementing the number step by step,
tracking which digits are present and when all digits from 0 to p-1 are seen.
"""

for _ in range(int(input())):
    n, p = map(int, input().split())
    a = list(map(int, input().split()))
    s = set(a)  # Set to track which digits are currently present
    L = a[-1]   # Last digit of the current number

    x = L
    # Find the first missing digit by traversing backwards
    while x in s:
        x = (x - 1 + p) % p  # Move to previous digit (wrapping around)
        if x == L:  # If we loop back to the start
            break

    # If we found a gap before the last digit
    if x < L:
        a[-1] = p - 1  # Set last digit to maximum value (p-1)
        # Propagate the carry
        for i in range(n - 1, -1, -1):
            s.add((a[i] + 1) % p)  # Add the next digit
            if (a[i] + 1) != p:    # If we didn't wrap around
                break
        else:
            s.add(1)  # Add 1 if we carried all the way to the front
        
        x = L  # Reset x
        # Find the first available digit after carry propagation
        while x in s:
            x = (x - 1 + p) % p
            if x == L or x == 0:  # Prevent unnecessary loop
                break

    # Output the number of operations needed
    print((x + p - L) % p)


# https://github.com/VaHiX/CodeForces/